### PR TITLE
Fixes for GA: hard param paths, SEO-related prerendering, 404 page, ...

### PR DESCRIPTION
**Pull request by harrysolovay**
This PR is quite comprehensive. It involves a rework of Capi & the frontend code pertaining to routing. We no longer utilize `stencil-route-link`. Instead, we highjack the routing ourselves, as it results in substantially-faster page changes (no need to reconcile 300+ alternative routes). Prerendering works solely on metadata, as Stencil's prerenderer proved volatile for our use case (although we're going to revisit this after GA). Some style refactoring and code cleanup as well.

### CODE REVIEW

File: tasks/build.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ const StencilBuildProcess = (flags: string[]) =>
   });
 
 const DEV_FLAGS = ["--dev", "--watch", "--serve"];
-const PROD_FLAGS: string[] = [];
+const PROD_FLAGS = ["--prerender", "--debug"];
```
Comment (jordanranz) [inline tasks/build.ts:29]: do we need debug here still?

File: package.json [hunk 2]
```diff
@@ -43,7 +43,8 @@
     "puppeteer": "1.20.0",
     "rollup-plugin-node-polyfills": "^0.2.1",
     "ts-node": "^8.5.0",
-    "typescript": "^3.7.5"
+    "typescript": "^3.8.3",
+    "workbox-build": "4.3.1"
```
Comment (jordanranz) [inline package.json:47]: what is `workbox-build`?

Comment (harrysolovay) [reply package.json:47]: Used by Stencil

File: tasks/build.ts (hunk 1 above)
Comment (harrysolovay) [reply tasks/build.ts:29]: No, but it doesn't hurt. Just incase we ever wanna debug console output

File: package.json (hunk 2 above)
Comment (jordanranz) [reply package.json:47]: Like installed by Stencil during the build?

File: client/stencil.config.ts [hunk 3]
```diff
@@ -1,15 +1,23 @@
 import {Config} from "@stencil/core";
 import nodePolyfills from "rollup-plugin-node-polyfills";
+import {sass} from "@stencil/sass";
 
 export const config: Config = {
-  globalScript: "src/global/global.ts",
+  // extras: {
```
Comment (jordanranz) [inline client/stencil.config.ts:6]: remove these commented lines

File: client/stencil.config.ts [hunk 4]
```diff
@@ -1,15 +1,23 @@
...
+  //   appendChildSlotFix: true,
+  //   slotChildNodesFix: true,
+  //   shadowDomShim: true,
+  // },
+  taskQueue: "async",
```
Comment (jordanranz) [inline client/stencil.config.ts:11]: We can probably get rid of this

File: client/src/styles/normalize.scss [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    background-color: rgba(0, 0, 0, 0.125);
+    cursor: pointer;
+  }
+
+  &::-webkit-scrollbar-thumb {
+    border-radius: 0;
+    background-color: rgba(0, 0, 0, 0.125);
+    cursor: pointer;
+  }
+*/
```
Comment (jordanranz) [inline client/src/styles/normalize.scss:14]: Can we remove this?

File: client/src/index.ts [hunk 6]
```diff
@@ -1,3 +1,9 @@
+// import {injectGlobalStyle} from "./amplify-ui/styles/styles";
+// import {injectFontFaceStyles} from "./global/inject-font-face-styles";
+
+// injectGlobalStyle();
+// injectFontFaceStyles();
+
```
Comment (jordanranz) [inline client/src/index.ts:6]: We can remove commented code

File: capi/src/visitors/links.ts [hunk 7]
```diff
@@ -26,19 +25,26 @@ const getRoute = (
...
   let {path: urlPath} = parsedURL;
 
   if (urlPath) {
-    if (urlPath.includes("?")) {
-      urlPath = urlPath.split("?").shift() as string;
+    if (urlPath.includes("#")) {
+      urlPath = urlPath.split("#").shift() as string;
+    }
+    let query = "";
+    if (urlPath.includes("/q/")) {
```
Comment (jordanranz) [inline capi/src/visitors/links.ts:36]: We need a check to make sure there is no page (`.md` file) named `q`

File: capi/src/visitors/links.ts [hunk 8]
```diff
@@ -26,19 +25,26 @@ const getRoute = (
...
+      const pieces = urlPath.split("/q/");
+      urlPath = pieces.shift() as string;
+      const paramEntry = (pieces.shift() as string).split("/");
+      query = `/q/${paramEntry[0]}/${paramEntry[1]}`;
```
Comment (jordanranz) [inline capi/src/visitors/links.ts:40]: We should probably add some checks around the variables here.

File: capi/src/write/filters-by-route.ts [hunk 9]
```diff
@@ -16,9 +16,13 @@ export async function filtersByRoute(
             ? ((): string | undefined => {
                 const filterKey = page?.filterKey as string | undefined;
                 return filterKey
-                  ? JSON.stringify({
-                      [filterKey]: ctx.config.filters[filterKey],
-                    })
+                  ? page.route.includes("/sdk")
```
Comment (jordanranz) [inline capi/src/write/filters-by-route.ts:19]: The chained ternary operator here can be cleaned up. Maybe move these into functions.

File: client/src/amplify-ui/toc/toc/toc.style.ts [hunk 10]
```diff
@@ -61,6 +61,7 @@ export const h3AnchorStyle = css`
 
 export const headerStyle = css`
   padding-left: 1.5rem;
+  padding-right: 1.5rem;
   padding-bottom: 0.5rem;
```
Comment (jordanranz) [inline client/src/amplify-ui/toc/toc/toc.style.ts:65]: not a blocker but this can be:
```
padding: 0 1.5rem 0.5rem 1.5rem;
```

Comment (jordanranz) [reply client/src/amplify-ui/toc/toc/toc.style.ts:65]: Also, these should be margins

File: capi/src/visitors/links.ts (hunk 7 above)
Comment (harrysolovay) [reply capi/src/visitors/links.ts:36]: Done

File: client/src/docs-ui/menu/menu.tsx [hunk 11]
```diff
@@ -16,84 +15,99 @@ export class DocsMenu {
...
-            subTitle: "(preview)",
-            href: "/lib",
-          }}
-        />
-      );
-    } else if (
-      this.page?.productRootLink?.route === "/ui" ||
-      this.page?.productRootLink?.route === "/ui-legacy"
-    ) {
-      if (this.selectedFilters?.framework !== "react-native") {
```
Comment (jordanranz) [inline client/src/docs-ui/menu/menu.tsx:50]: This line is missing now. We want to hide the switcher for react-native

File: client/stencil.config.ts [hunk 12]
```diff
@@ -1,15 +1,18 @@
 import {Config} from "@stencil/core";
 import nodePolyfills from "rollup-plugin-node-polyfills";
+import {sass} from "@stencil/sass";
 
 export const config: Config = {
-  globalScript: "src/global/global.ts",
+  taskQueue: "congestionAsync",
```
Comment (jordanranz) [inline client/stencil.config.ts:6]: this is the default so we can remove this

File: client/src/utils/ensure-menu-scrolled-into-view.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import {Breakpoint} from "../amplify-ui/styles/media";
+
+export const ensureMenuScrolledIntoView = (): void => {
+  const footer = document.querySelector("docs-footer");
+  const documentHeight = document.body.getBoundingClientRect().height;
+  if (footer && innerWidth <= Breakpoint.TABLET * 16) {
+    const footerHeight = footer.getBoundingClientRect().height;
+    if (scrollY > documentHeight - innerHeight - footerHeight) {
+      if (scrollY > documentHeight - innerHeight - footerHeight) {
```
Comment (jordanranz) [inline client/src/utils/ensure-menu-scrolled-into-view.ts:9]: Remove this

**Pull request merged**